metricbeat/module/windows/services/services_windows.go

```diff
@@ -0,0 +1,368 @@
+		if ServiceErrno(err.(syscall.Errno)) != SERVICE_ERROR_INSUFFICIENT_BUFFER {
+			if err := CloseServiceHandle(serviceHandle); err != nil {
+				return err
+			}
+			return err
+		}
+		serviceBufSize += serviceBytesNeeded
+		buffer := make([]byte, serviceBufSize)
+
+		for {
+			if err := _QueryServiceConfig(serviceHandle, &buffer[0], serviceBufSize, &serviceBytesNeeded); err != nil {
+				if ServiceErrno(err.(syscall.Errno)) != SERVICE_ERROR_INSUFFICIENT_BUFFER {
+					if err := CloseServiceHandle(serviceHandle); err != nil {
+						return err
+					}
+					return err
+				}
+				serviceBufSize += serviceBytesNeeded
+			} else {
+				serviceQueryConfig := (*QueryServiceConfig)(unsafe.Pointer(&buffer[0]))
+				service.StartType = serviceStartTypes[ServiceStartType(serviceQueryConfig.DwStartType)]
+				if err := CloseServiceHandle(serviceHandle); err != nil {
+					return err
+				}
+				break
+			}
+		}
+	}
+	return nil
+}
+
+func getServiceUptime(processId uint32) (gosigar.ProcTime, error) {
+	var processCreationTime gosigar.ProcTime
+
+	err := processCreationTime.Get(int(processId))
+	if err != nil {
+		return processCreationTime, err
+	}
+
+	return processCreationTime, nil
+}
+
+func (reader *ServiceReader) Close() error {
+	return CloseServiceDatabaseHandle(reader.handle)
+}
+
+func CloseServiceDatabaseHandle(handle ServiceDatabaseHandle) error {
+	if err := _CloseServiceHandle(uintptr(handle)); err != nil {
+		return ServiceErrno(err.(syscall.Errno))
+	}
+
+	return nil
+}
+
+func CloseServiceHandle(handle ServiceHandle) error {
+	if err := _CloseServiceHandle(uintptr(handle)); err != nil {
+		return ServiceErrno(err.(syscall.Errno))
+	}
+
+	return nil
+}
+
+func NewServiceReader() (*ServiceReader, error) {
+
+	hndl, err := OpenSCManager("", "", ScManagerEnumerateService|ScManagerConnect)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "initialization failed")
+	}
+
+	r := &ServiceReader{
+		handle: hndl,
+	}
+
+	//var state ServiceEnumState
+
+	// configState := strings.ToLower(config.State)
+	// switch configState {
+	// case "all", "":
+	// 	state = ServiceStateAll
+	// case "active":
+	// 	state = ServiceActive
+	// case "inactive":
+	// 	state = ServiceInActive
+	// default:
+	// 	err := fmt.Errorf("state '%s' are not valid", configState)
+	// 	r.Close()
+	// 	return nil, errors.Wrap(err, "initialization failed")
+	// }
+
+	r.state = ServiceStateAll
+
+	return r, nil
+}
+
+func (reader *ServiceReader) Read() ([]common.MapStr, error) {
+
+	services, err := getServiceStates(reader.handle, reader.state)
+
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]common.MapStr, 0, len(services))
+
+	for _, service := range services {
+		ev := common.MapStr{
+			"display_name": service.DisplayName,
+			"service_name": service.ServiceName,
+			"state":        service.CurrentState,
+			"start_type":   service.StartType,
+			"uptime":       service.Uptime,
```

Comment:
What are the units for this? I would embed them into the field name like `uptime.ms`. (This needs to be a nested object rather than a field name containing a dot.)